client/lib/tree-select/index.js

```diff
@@ -0,0 +1,69 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { isObject, some, isFunction } from 'lodash';
+
+/**
+ * Returns a selector that caches values.
+ *
+ * @param  {Function} getDependents A Function describing the dependent(s) of the selector.
+ *                                    Must return an array which gets passed as the first arg to the selector
+ * @param  {Function} selector      A standard selector for calculating cached result
+ * @return {Function}               Cached selector
+ */
+export default function treeSelect( getDependents, selector ) {
+	if ( process.env.NODE_ENV !== 'production' ) {
+		if ( ! isFunction( getDependents ) || ! isFunction( selector ) ) {
+			throw new TypeError(
+				'treeSelect: invalid arguments passed, selector and getDependents must both be functions'
+			);
+		}
+	}
+
+	const cache = new WeakMap();
+
+	const cachedSelector = function( state, ...args ) {
+		const dependents = getDependents( state, ...args );
+
+		if ( process.env.NODE_ENV !== 'production' ) {
+			if ( some( args, isObject ) ) {
+				throw new Error( 'Do not pass objects as arguments to a treeSelector' );
+			}
+		}
+
+		// create a dependency tree for caching selector results.
+		// this is beneficial over standard memoization techniques so that we can
+		// garbage collect any values that are based on outdated dependents
+		const cursor = dependents.reduce( insertDependentKey, cache );
+
+		const key = args.join();
+		if ( cursor.has( key ) ) {
+			return cursor.get( key );
+		}
+
+		const value = selector( dependents, ...args );
+		cursor.set( key, value );
+		return value;
+	};
+
+	return cachedSelector;
```

Comment:
why store it as a separate binding vs just returning the function expression?

Reply:
this is the nittiest of nits :)